include/marl/move_only_function.h

```diff
@@ -0,0 +1,194 @@
+// Copyright 2023 The Marl Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef marl_move_only_function_h
+#define marl_move_only_function_h
+
+#include "memory.h"
+
+#include <cstddef>
+#include <functional>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+namespace marl {
+
+#if __cplusplus > 201402L
+
+using std::invoke;
+
+#else
+// std::invoke for C++11 and C++14
+template <typename Functor, typename... Args>
+typename std::enable_if<
+    std::is_member_pointer<typename std::decay<Functor>::type>::value,
+    typename std::result_of<Functor && (Args && ...)>::type>::type
+invoke(Functor&& f, Args&&... args) {
+  return std::mem_fn(f)(std::forward<Args>(args)...);
+}
+
+template <typename Functor, typename... Args>
+typename std::enable_if<
+    !std::is_member_pointer<typename std::decay<Functor>::type>::value,
+    typename std::result_of<Functor && (Args && ...)>::type>::type
+invoke(Functor&& f, Args&&... args) {
+  return std::forward<Functor>(f)(std::forward<Args>(args)...);
+}
+
+#endif  // __cplusplus > 201402L
+
+template <class R>
+class move_only_function;
+
+#if __cplusplus > 201402L && (defined(__GNUC__) || defined(__clang__))
+template <class R, class... Args, bool kNoexcept>
+class move_only_function<R(Args...) noexcept(kNoexcept)> {
+#else
+template <class R, class... Args>
+class move_only_function<R(Args...)> {
+  static const bool kNoexcept = false;  // private
+#endif
+
+ public:
+  constexpr move_only_function() = default;
+  move_only_function(std::nullptr_t) {}
+
+  template <class F,
+            class = typename std::enable_if<
+                !std::is_same<typename std::decay<F>::type,
+                              move_only_function>::value>::type>
+  move_only_function(F&& function) {
+    if (sizeof(typename std::decay<F>::type) <= kMaximumOptimizableSize) {
```

Comment:
You'll need to check that the alignment is compatible for `std::decay<F>::type` too.